Sources/NIO/Codec.swift

```diff
@@ -101,115 +129,265 @@ private extension ChannelHandlerContext {
-        while var slice = self.cumulationBuffer?.slice(), slice.readableBytes > 0 {
-            // Needed to later calculate how much we need to advance the readerIndex of the cumulationBuffer.
-            let sliceReadable = slice.readableBytes
-            let sliceWriterIndex = slice.writerIndex
-
-            // We fetch the writerIndex of the cumulationBuffer to make a good guess about if the cumulationBuffer changed in between due re-entrant call to
-            // channelRead after we called decodeFunc(...).
-            let writerIndex = self.cumulationBuffer!.writerIndex
-            let result = try decodeFunc(ctx, &slice)
-
-            guard self.cumulationBuffer != nil else {
-                // The cumulationBuffer was set to nil by either removing the decoder or closing the channel, just break the loop.
-                break
-            }
+    /// Can we produce a buffer to be processed right now or not?
+    enum BufferAvailability {
+        /// No, because no bytes available
+        case nothingAvailable
+        /// No, because we're already processing one
+        case bufferAlreadyBeingProcessed
+        /// Yes please, here we go.
+        case available(ByteBuffer)
+    }
 
-            precondition(slice.writerIndex == sliceWriterIndex, "Writing to the buffer is not allowed")
+    /// Result of a try to process a buffer.
+    enum BufferProcessingResult {
+        /// Could not process a buffer because we are already processing one on the same call stack.
+        case cannotProcessReentrantly
+        /// Yes, we did process some.
+        case didProcess(DecodingState)
+    }
 
-            self.cumulationBuffer!.moveReaderIndex(forwardBy: sliceReadable - slice.readableBytes)
+    private var state: State = .ready
+    private var buffers: CircularBuffer<ByteBuffer> = CircularBuffer(initialRingCapacity: 4)
+}
 
-            // If the user told us more data is needed we also need to ensure the writerIndex did not change in between.
-            // If the writerIndex changed we need to retry as there is more data delivered via re-entrance maybe.
-            if result == .needMoreData && self.cumulationBuffer!.writerIndex == writerIndex {
-                break
+// MARK: B2MDBuffer Main API
+extension B2MDBuffer {
+    /// Start processing some bytes if possible, if we receive a returned buffer (through `.available(ByteBuffer)`)
+    /// we _must_ indicate the processing has finished by calling `finishProcessing`.
+    mutating func startProcessing() -> BufferAvailability {
+        switch self.state {
+        case .processingInProgress:
+            return .bufferAlreadyBeingProcessed
+        case .ready where self.buffers.count > 0:
+            var buffer = self.buffers.removeFirst()
+            buffer.writeBuffers(self.buffers)
+            self.buffers.removeAll(keepingCapacity: true)
```

Comment:
should we only keep the capacity up to a max value to minimise memory overhead ?